Modify cluster plan params through UI

mrsheepuk opened the pull request:
This PR addresses #489 allowing plan parameters to be specified when creating a cluster through the UI.

It exposes the parameters on cluster creation in a panel (hidden by default):

![image](https://user-images.githubusercontent.com/7505593/78892579-8df0b400-7a61-11ea-9722-48af8a6668b8.png)

When expanded, it shows the cluster parameters, by default only this which can be edited (with an option to show the read-only ones too):

![image](https://user-images.githubusercontent.com/7505593/78892661-ad87dc80-7a61-11ea-8eb5-cfb945327c2f.png)

The entered values are validated by the server and validation issues returned are shown alongside the relevant field:

![image](https://user-images.githubusercontent.com/7505593/78892739-d019f580-7a61-11ea-9e23-ff833c484dfe.png)

davet1985 commented on ui/lib/components/forms/ClusterBuildForm.js:
```diff
@@ -141,15 +146,37 @@ class ClusterBuildForm extends React.Component {
+  handlePlanOverride = planOverrides => {
+    this.setState({
+      planOverride: planOverrides
+    })
```
Need to ensure you're retaining the rest of the state.

davet1985 commented on ui/lib/components/plans/PlanOption.js:
```diff
@@ -0,0 +1,159 @@
+      return null
+    }
+
+    let displayName = this.props.displayName
+    if (!displayName) {
+      displayName = name
+    }
```
```suggestion
    const displayName = this.props.displayName || name
```

davet1985 commented on ui/lib/components/plans/PlanOption.js:
```diff
@@ -0,0 +1,159 @@
+    if (!displayName) {
+      displayName = name
+    }
+    let locked = null
+    if (!editable) locked = (
+      <Icon type="lock"/>
+    )
```
```suggestion
    const locked = !editable ? <Icon type="lock" /> : null
```

davet1985 commented on ui/lib/components/plans/PlanOption.js:
```diff
@@ -0,0 +1,159 @@
+    )
+
+    // Special handling for object types - represent as a card with a plan option for each property:
+    if (property.type == 'object') {
```
Should this be `===`?

mrsheepuk replied:
done

davet1985 commented on ui/lib/components/plans/PlanOption.js:
```diff
@@ -0,0 +1,159 @@
+
+    // Special handling for object types - represent as a card with a plan option for each property:
+    if (property.type == 'object') {
+      const keys = property.properties != null ? Object.keys(property.properties) : []
```
Should this be `!==`

mrsheepuk replied:
changed to just `property.properties ?` - null, undefined, falsy, false, whatever for this one.

davet1985 commented on ui/lib/components/plans/PlanOption.js:
```diff
@@ -0,0 +1,159 @@
+            return <InputNumber value={value} readOnly={!editable} disabled={!editable} onChange={(v) => onChange(name, v)} />
+          }
+          case 'array': {
+            const values = value != null ? value : []
+            if (property.items.type != 'array' && property.items.type != 'object') {
```
Should these be `!==`?

mrsheepuk replied:
done

davet1985 commented on ui/lib/components/plans/PlanOption.js:
```diff
@@ -0,0 +1,159 @@
+      <Form.Item label={startCase(displayName)} help={this.describe(property)}>
+        {(() => {
+          switch(property.type) {
+          case 'string': {
```
Should we indent the case blocks one more?

mrsheepuk replied:
lint failure if we do :(

davet1985 commented on ui/lib/components/plans/PlanOptionsForm.js:
```diff
@@ -0,0 +1,99 @@
+    this.setState({
+      planValues: newPlanValues
+    })
+    if (this.props.onPlanChange) this.props.onPlanChange(newPlanValues)
```
Braces 😂

mrsheepuk replied:
ffs

davet1985 replied:
I'll rebase that linting PR soon

mrsheepuk replied:
done

davet1985 commented on ui/lib/components/plans/PlanOptionsForm.js:
```diff
@@ -0,0 +1,99 @@
+    const newPlanValues = set({...this.state.planValues}, name, value)
+    this.setState({
+      planValues: newPlanValues
+    })
```
Need to ensure we are retaining the rest of the state object here?

davet1985 commented on ui/lib/components/plans/PlanOptionsForm.js:
```diff
@@ -0,0 +1,99 @@
+  handleShowReadOnlyChange = (e) => {
+    this.setState({
+      showReadOnly: e.target.checked
+    })
```
And again, state retention?

davet1985 commented on ui/lib/components/plans/PlanOptionsForm.js:
```diff
@@ -0,0 +1,99 @@
+    }
+    return (
+      <Collapse>
+        <Collapse.Panel header="Customize cluster parameters">
```
American spelling?

mrsheepuk replied:
following @gambol99 's lead from the CLI/code/docs... i'd prefer British English myself...

davet1985 replied:
Oh right, so we can't actually use `customise`?!

mrsheepuk replied:
leaving as per discussion in slack. one to re-open with a wider i18n piece perhaps if we ever get that far.

gambol99 commented:
In relation to the API bits .. it's an LGTM from me ..

davet1985 commented:
Maybe a little more space around the validation errors?

<img width="1178" alt="Screen Shot 2020-04-09 at 14 13 47" src="https://user-images.githubusercontent.com/1334068/78898848-618e6500-7a6c-11ea-869e-82ec78f79cc3.png">

davet1985 commented:
I've tested as follows:
- creating a cluster with no modifications to the plan
- making incorrect modifications to the plan and receiving in-form error messages
- creating a cluster with modifications to the plan, such as Auth Proxy Allowed IPs, size and version

This all worked as expected, I saw the modified plan parameters appearing on the cluster resource. The only question I have is, I created a cluster with an overridden Auth Proxy Allowed IPs set to my home IP, but I can get access to the cluster (through kubectl) on my home IP and other IPs. I didn’t think I should be able to?

mrsheepuk commented:
> Maybe a little more space around the validation errors?
> 
> <img alt="Screen Shot 2020-04-09 at 14 13 47" width="1178" src="https://user-images.githubusercontent.com/1334068/78898848-618e6500-7a6c-11ea-869e-82ec78f79cc3.png">

done, added 10px spacing, looks better now. Rebased onto master and all ready to go.
